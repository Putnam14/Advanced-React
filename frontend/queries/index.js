import gql from 'graphql-tag';
import { itemDetails } from './fragments';

export const CREATE_ITEM_MUTATION = gql`
  ${itemDetails}
  mutation createItem($description: String!, $title: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(description: $description, title: $title, price: $price, image: $image, largeImage: $largeImage) {
      ...itemDetails
    }
  }
`;

export const SIGNUP_MUTATION = gql`
  mutation signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const SIGNIN_MUTATION = gql`
  mutation signup($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const REQUEST_RESET_MUTATION = gql`
  mutation requestReset($email: String!) {
    requestReset(email: $email) {
      id
    }
  }
`;

export const RESET_MUTATION = gql`
  mutation resetPassword($resetToken: String!, $password: String!, $confirmPassword: String!) {
    resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const CREATE_ORDER_MUTATION = gql`
  mutation CreateOrderMutation($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

export const ALL_ITEMS_QUERY = gql`
  # Import the Fragment
  ${itemDetails}
  query AllItemsQuery($skip: Int = 0, $first: Int = 2) {
    itemsConnection(orderBy: createdAt_DESC, first: $first, skip: $skip) @connection(key: "itemsConnection") {
      aggregate {
        count
      }
    }
    items(orderBy: createdAt_DESC, first: $first, skip: $skip) @connection(key: "items") {
      ...itemDetails
    }
  }
`;

export const SINGLE_ITEM_QUERY = gql`
  ${itemDetails}
  query FindItem($id: ID!) {
    items(where: { id: $id }) {
      user {
        id
        email
        name
      }
      ...itemDetails
    }
  }
`;

export const SINGLE_ORDER_QUERY = gql`
  query order($id: ID!) {
    order(where: { id: $id }) {
      id
      charge
      total
      createdAt
      items {
        id
        title
        price
        description
        image
        quantity
      }
    }
  }
`;

export const SEARCH_ITEMS_QUERY = gql`
  ${itemDetails}
  query SearchItems($searchTerm: String!) {
    items(
      where: {
        OR: [
          { title_contains: $searchTerm }
          { description_contains: $searchTerm }
        ]
      }
    ) {
      ...itemDetails
    }
  }
`;

export const REMOVE_ITEM_MUTATION = gql`
  mutation RemoveItem($id: ID!) {
    deleteItem(id: $id) {
      id
      title
      description
    }
  }
`;

export const UPDATE_ITEM_MUTATION = gql`
  ${itemDetails}
  mutation UpdateItem(
    $id: ID!
    $title: String
    $description: String
    $price: Int
  ) {
    updateItem(
      id: $id
      description: $description
      title: $title
      price: $price
    ) {
      ...itemDetails
    }
  }
`;

export const UPDATE_USER_MUTATION = gql`
  mutation UpdateUser($name: String!) {
    updateUser(name: $name) {
      name
    }
  }
`;

export const CURRENT_USER_QUERY = gql`
  ${itemDetails}
  query userQuery {
    me {
      id
      email
      name
      orders {
        charge
        id
        total
      }
      cart {
        id
        quantity
        item {
          ...itemDetails
        }
      }
    }
  }
`;

export const USER_ORDERS_QUERY = gql`
  query orders {
    orders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      updatedAt
      items {
        id
        title
        price
        description
        quantity
        image
      }
    }
  }
`;

export const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
      item {
        id
        price
        description
        image
        title
      }
    }
  }
`;

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

export const ALL_USERS_QUERY = gql`
  query Users {
    users {
      id
      name
      email
      permissions
    }
  }
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation updatePermissions($permissions: [Permission], $userId: ID!) {
    updatePermissions(permissions: $permissions, userId: $userId) {
      id
      permissions
      name
      email
    }
  }
`;
