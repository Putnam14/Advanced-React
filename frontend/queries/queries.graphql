fragment itemDetails on Item {
  __typename
  id
  title
  price
  description
  image
  largeImage
}

mutation CREATE_ITEM_MUTATION($description: String!, $title: String!, $price: Int!, $image: String, $largeImage: String) {
  createItem(description: $description, title: $title, price: $price, image: $image, largeImage: $largeImage) {
    ...itemDetails
  }
}

mutation SIGNUP_MUTATION($email: String!, $name: String!, $password: String!) {
  signup(email: $email, name: $name, password: $password) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation SIGNIN_MUTATION($email: String!, $password: String!) {
  signin(email: $email, password: $password) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation REQUEST_RESET_MUTATION($email: String!) {
  requestReset(email: $email) {
    id
  }
}

mutation RESET_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
  resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation CREATE_ORDER_MUTATION($token: String!) {
  createOrder(token: $token) {
    id
    charge
    total
    items {
      id
      title
    }
  }
}

query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
  itemsConnection(orderBy: createdAt_DESC, first: $first, skip: $skip) {
    aggregate {
      count
    }
  }
  items(orderBy: createdAt_DESC, first: $first, skip: $skip) {
    ...itemDetails
  }
}

query SINGLE_ITEM_QUERY($id: ID!) {
  items(where: { id: $id }) {
    user {
      id
      email
      name
    }
    ...itemDetails
  }
}

query SINGLE_ORDER_QUERY($id: ID!) {
  order(id: $id) {
    id
    charge
    total
    createdAt
    user {
      id
    }
    items {
      id
      title
      price
      description
      image
      quantity
    }
  }
}

query SEARCH_ITEMS_QUERY($searchTerm: String!) {
  items(
    where: {
      OR: [
        { title_contains: $searchTerm }
        { description_contains: $searchTerm }
      ]
    }
  ) {
    ...itemDetails
  }
}

mutation REMOVE_ITEM_MUTATION($id: ID!) {
  deleteItem(id: $id) {
    id
    title
    description
  }
}

mutation UPDATE_ITEM_MUTATION(
  $id: ID!
  $title: String
  $description: String
  $price: Int
) {
  updateItem(
    id: $id
    description: $description
    title: $title
    price: $price
  ) {
    ...itemDetails
  }
}

mutation UPDATE_USER_MUTATION($name: String!) {
  updateUser(name: $name) {
    name
  }
}

query CURRENT_USER_QUERY {
  me {
    id
    email
    name
    permissions
    orders {
      id
      charge
      total
    }
    cart {
      id
      quantity
      item {
        ...itemDetails
      }
    }
  }
}

query USER_ORDERS_QUERY {
  orders(orderBy: createdAt_DESC) {
    id
    total
    createdAt
    updatedAt
    items {
      id
      title
      price
      description
      quantity
      image
    }
  }
}

mutation ADD_TO_CART_MUTATION($id: ID!) {
  addToCart(id: $id) {
    id
    quantity
    item {
      id
      price
      description
      image
      title
    }
  }
}

mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
  removeFromCart(id: $id) {
    id
  }
}

query ALL_USERS_QUERY {
  users {
    id
    name
    email
    permissions
  }
}

mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userId: ID!) {
  updatePermissions(permissions: $permissions, userId: $userId) {
    id
    permissions
    name
    email
  }
}

query LOCAL_STATE_QUERY {
  cartOpen @client
}

mutation TOGGLE_CART_MUTATION {
  toggleCart @client
}

