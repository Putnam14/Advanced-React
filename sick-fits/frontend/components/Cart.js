import React from 'react'
import { Query, Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import { adopt } from 'react-adopt'
import CartStyles from './styles/CartStyles'
import Supreme from './styles/Supreme'
import CloseButton from './styles/CloseButton'
import SickButton from './styles/SickButton'
import User from './User'
import CartItem from './CartItem'
import calcTotalPrice from '../lib/calcTotalPrice'
import formatMoney from '../lib/formatMoney'

const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`
/* eslint-disable */
const Composed = adopt({
  // the render arrow function gets rid of complaints about no children
  // could also do 'user: <User />,'
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
})
/* eslint-enable */

const Cart = () => (
  <Composed>
    {({ user, toggleCart, localState }) => {
      const { me } = user.data
      if (!me) return null
      return (
        <CartStyles open={localState.data.cartOpen}>
          <header>
            <CloseButton title="close" onClick={toggleCart}>
              &times;
            </CloseButton>
            <Supreme>
              {me.name}
              's Cart
            </Supreme>
            <p>
              You have {me.cart.length} item
              {me.cart.length >= 2 ? 's' : ''} in your cart
            </p>
          </header>
          <ul>
            {me.cart.map(cartItem => (
              <CartItem key={cartItem.id} cartItem={cartItem} />
            ))}
          </ul>
          <footer>
            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
            <SickButton>Checkout</SickButton>
          </footer>
        </CartStyles>
      )
    }}
  </Composed>
)

export default Cart
export { LOCAL_STATE_QUERY }
export { TOGGLE_CART_MUTATION }
