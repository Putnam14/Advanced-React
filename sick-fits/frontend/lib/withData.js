// Next-Apollo package to make sure server-side rendering works
import withApollo from 'next-with-apollo'
// You can import Apollo along with any 'link's you want, or do this
import ApolloClient from 'apollo-boost'
import { endpoint } from '../config'

function createClient({ headers }) {
  return new ApolloClient({
    // Can change this if you have different Yoga APIs for dev, prod
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          // Bring cookies along for the ride
          credentials: 'include',
        },
        headers,
      })
    },
  })
}

// Use to create client in _app.js
export default withApollo(createClient)
