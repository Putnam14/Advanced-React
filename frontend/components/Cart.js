import React from 'react';
import { Mutation, Query } from 'react-apollo';
import { adopt } from 'react-adopt';
import TakeMyMoney from './TakeMyMoney';
import formatMoney from '../lib/formatMoney';
import CartItem from './CartItem';
import { CURRENT_USER_QUERY, LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION } from '../queries/queries';
import calcTotalPrice from '../lib/calcTotalPrice';
import Error from './ErrorMessage';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';

const Composed = adopt({
  toggleCart: <Mutation mutation={TOGGLE_CART_MUTATION}>{() => {}}</Mutation>,
  localState: <Query query={LOCAL_STATE_QUERY}>{() => {}}</Query>,
  currentUser: (
    <Query query={CURRENT_USER_QUERY} data-test="cart">
      {() => {}}
    </Query>
  ),
});

const Cart = () => (
  <Composed>
    {({ toggleCart, localState, currentUser }) => {
      const { data: { me }, error, loading } = currentUser;
      if (loading) return <p>Loading...</p>;
      if (error) return <Error error={error} />;
      if (!me) return <p>Please Sign In!</p>;
      return (
        <CartStyles open={localState.data.cartOpen}>
          <header>
            <CloseButton title="close" onClick={toggleCart}>
              &times;
            </CloseButton>

            <Supreme>{me.name}'s Cart.</Supreme>
            <p>
              You have {me.cart.length} item{me.cart.length === 1 ? '' : 's'} in your cart.
            </p>
          </header>

          <ul>{me.cart.map(cartItem => <CartItem key={cartItem.id} cartItem={cartItem} />)}</ul>
          <footer>
            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
            <TakeMyMoney>
              <SickButton>Checkout</SickButton>
            </TakeMyMoney>
          </footer>
        </CartStyles>
      );
    }}
  </Composed>
);

export default Cart;
