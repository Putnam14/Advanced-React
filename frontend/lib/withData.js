import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from '../queries/queries.graphql';
// can also be a function that accepts a `headers` object (SSR only) and returns a config

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? 'http://localhost:4444' : 'http://localhost:4444',
    ssrMode: !process.browser, // Disables forceFetch on the server (so queries are only run once)
    request: async operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            const { cartOpen } = cache.read({ query: LOCAL_STATE_QUERY });
            const data = { data: { cartOpen: !cartOpen } };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo({ client: createClient });
