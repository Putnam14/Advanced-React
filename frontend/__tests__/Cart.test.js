import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import mountOptions from '../lib/testUtils';
import wait from 'waait';
import Cart from '../components/Cart';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeCartItem, fakeUser } from '../lib/testUtils';
import { CURRENT_USER_QUERY, LOCAL_STATE_QUERY } from '../queries/queries';

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem()],
        },
      },
    },
  },
  {
    request: { query: LOCAL_STATE_QUERY },
    result: {
      data: {
        cartOpen: true,
      },
    },
  },
];

describe('<Cart/>', () => {
  it('renders', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Cart />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('header'))).toMatchSnapshot();
    expect(wrapper.find('CartItem')).toHaveLength(1);
  });
});
