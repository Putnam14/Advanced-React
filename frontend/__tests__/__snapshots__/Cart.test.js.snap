// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart/> renders 1`] = `
<header>
  <styled.button
    onClick={[Function]}
    title="close"
  >
    <button
      className="sc-EHOje jytiuo"
      onClick={[Function]}
      title="close"
    >
      Ã—
    </button>
  </styled.button>
  <styled.h3>
    <h3
      className="sc-ifAKCX imryuh"
    >
      Miss Coleman Berge
      's Cart.
    </h3>
  </styled.h3>
  <p>
    You have 
    1
     item
     in your cart.
  </p>
</header>
`;
